libr/debug/p/debug_native.c

```diff
@@ -273,84 +285,107 @@ static int r_debug_native_wait (RDebug *dbg, int pid) {
-#if __linux__ && defined(PT_GETEVENTMSG)
-			// Handle PTRACE_EVENT_*
-			if (WIFSTOPPED (status) && WSTOPSIG (status) == SIGTRAP) {
-				ut32 pt_evt = status >> 16;
-				ut32 data;
-
-				switch (pt_evt) {
-				case 0:
-					// Normal trap?
-					break;
-
-				case PTRACE_EVENT_FORK:
-					if (dbg->trace_forks) {
-						if (ptrace (PTRACE_GETEVENTMSG, pid, 0, &data) == -1) {
-							r_sys_perror ("ptrace GETEVENTMSG");
-						} else {
-							eprintf ("PTRACE_EVENT_FORK new_pid=%d\n", data);
-							dbg->forked_pid = data;
-							// TODO: more handling here?
-						}
-					}
-					break;
-				case PTRACE_EVENT_EXIT:
-					if (ptrace (PTRACE_GETEVENTMSG, pid, 0, &data) == -1) {
-						r_sys_perror ("ptrace GETEVENTMSG");
-					} else {
-						eprintf ("PTRACE_EVENT_EXIT pid=%d, status=%d\n", pid, data);
-					}
-					break;
-				default:
-					eprintf ("Unknown PTRACE_EVENT encountered: %d\n", pt_evt);
-					break;
-				}
-			}
-#endif
+	//eprintf ("r_debug_native_wait: status=%d (0x%x) (return=%d)\n", status, status, ret);
 
-			r_debug_handle_signals (dbg);
+#ifdef WAIT_ON_ALL_CHILDREN
+	if (ret != pid) {
+		reason = R_DEBUG_REASON_NEW_PID;
+		eprintf ("switching to pid %d\n", ret);
+		r_debug_select(dbg, ret, ret);
+	}
+#endif // WAIT_ON_ALL_CHILDREN
 
-			if (WIFSTOPPED (status)) {
-				dbg->reason.signum = WSTOPSIG (status);
-				status = R_DEBUG_REASON_SIGNAL;
-			} else if (status == 0 || ret == -1) {
-				status = R_DEBUG_REASON_DEAD;
-			} else {
-				if (ret != pid)
-					status = R_DEBUG_REASON_NEW_PID;
-				else status = R_DEBUG_REASON_UNKNOWN;
-			}
+	// TODO: switch status and handle reasons here
+#if __linux__ && defined(PT_GETEVENTMSG)
+	reason = linux_ptrace_event (dbg, pid, status);
+#endif // __linux__
+
+	/* propagate errors */
+	if (reason == R_DEBUG_REASON_ERROR)
+		return reason;
+
+	/* we don't know what to do yet, let's try harder to figure it out. */
+	if (reason == R_DEBUG_REASON_UNKNOWN) {
+		if (WIFEXITED (status)) {
+			eprintf ("child exited with status %d\n", WEXITSTATUS(status));
+			reason = R_DEBUG_REASON_DEAD;
+		} else if (WIFSIGNALED (status)) {
+			eprintf ("child received signal %d\n", WTERMSIG(status));
+			reason = R_DEBUG_REASON_SIGNAL;
+		} else if (WIFSTOPPED (status)) {
+			eprintf ("child stopped with signal %d\n", WSTOPSIG (status));
+
+			/* this one might be good enough... */
+			dbg->reason.signum = WSTOPSIG (status);
+
+			/* the ptrace documentation says GETSIGINFO is only necessary for
+			 * differentiating the various stops.
+			 *
+			 * this might modify dbg->reason.signum
+			 */
+			if (!r_debug_handle_signals (dbg))
+				return R_DEBUG_REASON_ERROR;
+			reason = dbg->reason.type;
+		} else if (WIFCONTINUED (status)) {
+			eprintf ("child continued...\n");
+			reason = R_DEBUG_REASON_NONE;
+		} else if (status == 1) {
+			/* XXX(jjd): does this actually happen? */
+			eprintf ("EEK DEAD DEBUGEE!\n");
+			reason = R_DEBUG_REASON_DEAD;
+		} else if (status == 0) {
```

Comment:
Seems like it came from 233982c965f0f077a5241254e0b5bd0b13e0a27b and has something to do with dead inferiors. No clue on platform...